packages/widget-v2/src/modals/TokenAndChainSelectorModal/useFilteredChains.ts

```diff
@@ -0,0 +1,68 @@
+import { matchSorter } from "match-sorter";
+import { useMemo } from "react";
+import { ChainWithAsset, GroupedAsset } from "./TokenAndChainSelectorModal";
+import { skipChainsAtom } from "@/state/skipClient";
+import { useAtomValue } from "jotai";
+import { useGetBalance } from "@/hooks/useGetBalance";
+import { chainFilterAtom } from "@/state/swapPage";
+
+export type useFilteredChainsProps = {
+  selectedGroup: GroupedAsset | undefined;
+  searchQuery: string;
+  context: "source" | "destination";
+};
+
+export const useFilteredChains = ({
+  selectedGroup,
+  searchQuery,
+  context,
+}: useFilteredChainsProps) => {
+  const { data: chains } = useAtomValue(skipChainsAtom);
+  const chainFilter = useAtomValue(chainFilterAtom);
+  const getBalance = useGetBalance();
+
+  const filteredChains = useMemo(() => {
+    if (!selectedGroup || !chains) return;
+
+    const chainsWithAssets = selectedGroup.assets
+      .map((asset) => {
+        const chain = chains.find((c) => c.chainID === asset.chainID);
+        return chain ? { ...chain, asset } : null;
+      })
+      .filter((chain) => {
+        if (!chain) return false;
+        // Check if chain is allowed based on chainFilter
+        const isAllowedByFilter = !chainFilter?.[context] ||
+          Object.keys(chainFilter[context]).includes(chain.chainID);
+
+        // For source context, exclude Penumbra chains
+        const isPenumbraAllowed = context !== "source" ||
+          !chain.chainID.includes("penumbra");
+
+        return isAllowedByFilter && isPenumbraAllowed;
+      }) as ChainWithAsset[];
```

Comment:
filtering by `filter` prop implemented here for chains, also moved filtering out penumbra from source chains here